cin_validator/rules/cin2022_23/rule_8930.py

```diff
@@ -0,0 +1,107 @@
+from cin_validator.utils import make_census_period
+
+# Get tables and columns of interest from the CINTable object defined in rule_engine/__api.py
+# Replace ChildIdentifiers with the table name, and LAChildID with the column name you want.
+
+ChildProtectionPlans = CINTable.ChildProtectionPlans
+CPPendDate = ChildProtectionPlans.CPPendDate
+Header = CINTable.Header
+ReferenceDate = Header.ReferenceDate
+
+# define characteristics of rule
+@rule_definition(
+    # write the rule code here, in place of 8930
+    code=8930,
+    # replace ChildIdentifiers with the value in the module column of the excel sheet corresponding to this rule .
+    module=CINTable.ChildProtectionPlans,
+    # replace the message with the corresponding value for this rule, gotten from the excel sheet.
+    message="Child Protection Plan End Date must fall within the census year",
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[CPPendDate, ReferenceDate],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # Replace ChildIdentifiers with the name of the table you need.
+    df = data_container[ChildProtectionPlans]
+
+    # ReferenceDate exists in the heder table so we get header table too.
+    df_ref = data_container[Header]
+    ref_date_series = df_ref[ReferenceDate]
+
+    # the make_census_period function generates the start and end date so that you don't have to do it each time.
+    collection_start, reference_date = make_census_period(ref_date_series)
+
+    # implement rule logic as described by the Github issue. Put the description as a comment above the implementation as shown.
+
+    # If <CPPendDate> (N00115) is present, then<CPPendDate> (N00115) must fall within [Period_of_Census] inclusive
+    failing_indices = df[
+        (df[CPPendDate] < collection_start) | (df[CPPendDate] > reference_date)
+    ].index
+
+    # Replace ChildIdentifiers and LAchildID with the table and column name concerned in your rule, respectively.
+    # If there are multiple columns or table, make this sentence multiple times.
+    rule_context.push_issue(
+        table=ChildProtectionPlans, field=CPPendDate, row=failing_indices
+    )
+
+
+def test_validate():
+    # Create some sample data such that some values pass the validation and some fail.
+    fake_header = pd.DataFrame(
+        [{ReferenceDate: "31/03/2022"}]  # the census start date here will be 01/04/2021
+    )
+
+    fake_cpp = pd.DataFrame(
+        [
+            {
+                CPPendDate: "01/03/2019"
+            },  # 0 fail: March 1st is before April 1st, 2021. It is out of range
+            {
+                CPPendDate: "01/04/2021"
+            },  # 1 pass: April 1st is within April 1st, 2021 to March 31st, 2022.
+            {
+                CPPendDate: "01/10/2022"
+            },  # 2 fail: October 1st is after March 31st, 2022. It is out of range
```

Comment:
You've checked for when the date is before, within and after the range.
Good test coverage. 

What if the date is not present at all?
You can a row where CPPendDate is pd.NA (that is, null)

Reply:
When you add the row, run the rule and see what happens. Does the row with the missing value fail, pass or is it ignored?. @amyrichards0n

Reply:
On second thought, let's do this in the future.

Since it is a little change compared to the work you've already done, I can make the change and merge it in.
See you tomorrow!.